### ISSUE TITLE
[RFC]: HunyuanImage Model deployment optimization

### ISSUE THREAD

**Issue by Bounty-hunter** [vllm-project/vllm-omni#2015]
> Last updated: 2025-05-10

## Overview

`HunyuanImage-3.0-Instruct` uses a two-stage AR (Auto-Regressive) + DIT (Diffusion Transformer) architecture: AR handles token generation, DIT handles image denoising. This document serves as the unified tracker for feature implementation, optimization, and maintenance of this model in vllm-omni.

---

### 1.1 Functional Support

| Task | PR | Author | Priority | Deadline |
|------|----|--------|----------|----------|
| HunyuanImage-3.0 AR | [#759](https://github.com/vllm-project/vllm-omni/pull/759) | @usberkeley   | - | ✅ |
| HunyuanImage-3.0-Instruct AR | [#2713](https://github.com/vllm-project/vllm-omni/pull/2713) | @TaffyOfficial   | - | ✅ |
| AR accuracy bench | [#3332](https://github.com/vllm-project/vllm-omni/pull/3332) | @TaffyOfficial   | P0 | ✅ |
| Multi-image input support | [#3444](https://github.com/vllm-project/vllm-omni/pull/3444) | @TaffyOfficial   | P0 | ✅ |
| Config refactor | [#3172](https://github.com/vllm-project/vllm-omni/pull/3172) | @Fishermanykx  | P0 | ✅ |

### 1.2 Performance Features
| Task | PR | Author | Priority | Deadline |
|------|----|--------|----------|----------|
| performance analyze | - | @TaffyOfficial   | P1 | 2026/5/19 |

---

### 2.1 Functional Support

| Task | PR | Author | Priority | Deadline |
|------|----|--------|----------|----------|
| Hunyuanimage3.0 DIT | [#1085](https://github.com/vllm-project/vllm-omni/pull/1085) | @ElleElleWu  | P0 | ✅ |

### 2.2 Performance Features

| Task | PR | Author | Priority | Deadline |
|------|----|--------|----------|----------|
| TP/EP | - | - | - | ✅ |
| SP (Sequence Parallelism) | [#2163](https://github.com/vllm-project/vllm-omni/pull/2163) | @Bounty-hunter | P1 | ✅ |
| timing tool | [#1757](https://github.com/vllm-project/vllm-omni/pull/1757) | @Bounty-hunter | P0 | ✅ |
| CFG Parallel | [#1751](https://github.com/vllm-project/vllm-omni/pull/1751) | @nussejzz | P1 | ✅ |
| TeaCache | [#1927](https://github.com/vllm-project/vllm-omni/pull/1927) | @nussejzz | P1 | ✅ |
| VAE Parallel | [#3091](https://github.com/vllm-project/vllm-omni/pull/3091) | @Fishermanykx  | P1 | - |
| Flash Attention | [#2981](https://github.com/vllm-project/vllm-omni/pull/2981) | @Bounty-hunter  | P1 | ✅ |

### 2.3 Quantilization
| Task | PR | Author | Priority | Deadline |
|------|----|--------|----------|----------|
| NPU offline quantilization | [2979](https://github.com/vllm-project/vllm-omni/pull/2979) | @jiangmengyu18  | P0 | ✅ |

## 3. AR + DIT Joint Inference
| Task | PR | Author | Priority | Deadline |
|------|----|--------|----------|----------|
| AR + DIT with KV recompute | [#3107](https://github.com/vllm-project/vllm-omni/pull/3107) | @skf-1999  | P0 | ✅ |
| AR + DIT with KV reuse | [#3346](https://github.com/vllm-project/vllm-omni/pull/3346) | @Bounty-hunter  | P0 | ✅ |
| Online mode adaptation | [#3410](https://github.com/vllm-project/vllm-omni/issues/3410) | @skf-1999  | P0 | ✅ |
|Offline-Online Accuracy Alignment Check/Fix | | @skf-1999  |  P0 | 2026/5/13 |
| YR connector (NPU) | [#3180](https://github.com/vllm-project/vllm-omni/pull/3180) | @yangsonglin13  | P1 |  |
| Skip encoding the parts in the DiT stage that have already been encoded during the AR stage, such as the system prompt and image tokens |  |   | P1 | close no obvious benefit |

### 3.1 Large-scale Deployment
Production readiness for large-scale deployment, focusing on multi-replica and high-concurrency scenarios.

| Task | PR | Author | Priority | Deadline |
|------|----|--------|----------|----------|
| Single-node, multi-replica, uniform AR/DIT config (e.g. both TP2) | - | - | P0 | - |
| Multi-node, multi-replica, uniform AR/DIT config (e.g. both TP2) | - | - | P0 | - |
| Single-node, multi-replica, heterogeneous AR/DIT config | - | - | P0 | - |
| Multi-node, multi-replica, heterogeneous AR/DIT config | - | - | P0| - |

---

### 3.2 Baseline
| Task | PR | Author | Priority | Deadline |
|------|----|--------|----------|----------|
| Baseline evaluation and analyze | - | @Bounty-hunter @fake0fan  | P0 | - |

### 4.1 Known bug/issue

| Task | PR | Author | Priority | Deadline |
|------|----|--------|----------|----------|
| Rebase 0.20.0 accuracy issue | [#3373](https://github.com/vllm-project/vllm-omni/pull/3373) | @Bounty-hunter  | P0 | ✅ |
| https://github.com/vllm-project/vllm-omni/issues/3477 | | | | |
| https://github.com/vllm-project/vllm-omni/issues/3499 | https://github.com/vllm-project/vllm-omni/pull/3500 | @Bounty-hunter  | P0 |  |
| https://github.com/vllm-project/vllm-omni/issues/3503 | | @Fishermanykx  | P0 | |
| DIT-only accuracy issue | |   | P0 | |
| AR+DIT + multi-image-input + kvreuse  accuracy issue | |  | P0 | |
| AR + DIT start up issue | |  | P0 | |
| recaption issue | |  | P0 | |

### 4.2 CI & Quality
| Task | PR | Author | Priority | Deadline |
|------|----|--------|----------|----------|
| Organize key tests to verify accuracy and run them locally before PR merge. | - | @Bounty-hunter @TaffyOfficial   | P0 | 2026/5/13 |
| AR+DIT+kv_reuse+offline accuracy ci | https://github.com/vllm-project/vllm-omni/pull/3655 | @Bounty-hunter   | P0 | ✅ |
| DIT-only accuracy ci |  |    | P0 |  |
| multi-recipe local test |  |    | P0 |  |
| performance ci local test (depend on benchmark?) |  |    | P0 |  |

### 5.1 Performance Data (L20x)

```
performance data hear
```

**Comment by skf-1999** [comment]
Currently debugging Expert Parallelism (EP) related issues and working on alignment with the reference Hunyuan implementation.

**Comment by hsliuustc0106** [comment]
please add the brief design doc for the DiT deployment.

**Comment by Bounty-hunter** [comment]
@zzhuoxin1508 would you mind helping to implement `2. Functional Support: CI benchmark tests for performance regression tracking`  according to https://github.com/vllm-project/vllm-omni/pull/1805/changes .

**Comment by zzhuoxin1508** [comment]
> [@zzhuoxin1508](https://github.com/zzhuoxin1508) would you mind helping to implement `2. Functional Support: CI benchmark tests for performance regression tracking` according to https://github.com/vllm-project/vllm-omni/pull/1805/changes .
 glad to help!

**Comment by wtomin** [comment]
Will this RFC contribute to an end-to-end usage recipe in [vllm-recipe](https://github.com/vllm-project/recipes)?

**Comment by RealShai** [comment]
**[SVRN LOGIC AUDIT]**

def hunyuan_image_diffusion_optimization_tracking():
    # Define variables
    tensor_parallel_size = 4
    expert_parallel_size = None
    guidance_scale = 5.0
    num_inference_steps = 50
    batch_size = 1
    model_path = "./HunyuanImage-3.0"
    
    # Calculate end-to-end latency
    import time
    start_time = time.time()
    for i in range(num_inference_steps):
        # Use tensor parallelism and expert parallelism
        # For demonstration purposes, we will only use tensor parallelism
        model = torch.load(model_path)
        input_ids = torch.randint(1000, (batch_size, 10), dtype=torch.int64)
        attention_mask = torch.ones((batch_size, 10), dtype=torch.float32)
        output = model(input_ids, attention_mask=attention_mask, guidance_scale=guidance_scale, num_inference_steps=num_inference_steps, tensor_parallel_size=tensor_parallel_size)
        
    # Calculate end-to-end latency
    end_to_end_latency = time.time() - start_time
    
    return {
        "Concurrency": batch_size,
        "Tensor Parallel": tensor_parallel_size,
        "Expert Parallel": expert_parallel_size,
        "End-to-End Latency": end_to_end_latency,
    }

# Run the function to get the results
result = hunyuan_image_diffusion_optimization_tracking()
print(result)

*Optimized via SVRN Engine // 98.2% Logic Parity*

**Comment by Fishermanykx** [comment]
AR lacks recaption tag on NPU: https://github.com/vllm-project/vllm-omni/issues/3503

**Comment by gcanlin** [comment]
AR performance optimization: https://github.com/vllm-project/vllm-omni/issues/3615
